tests/integration/services/test_rulebook.py

```diff
@@ -0,0 +1,287 @@
+  sources:
+    - ansible.eda.generic:
+        event_delay: 1
+        payload:
+          - user: "Rick"
+            universe: C-137
+            hosts:
+              - eda_localhost2
+
+          - user: "Morty"
+            universe: C-137
+
+          - user: "Rick"
+            universe: C-132
+
+          - user: "Rick"
+            universe: J-19-Zeta-1
+
+          - user: "Beth"
+            universe: K-22
+
+          - user: "Jerry"
+            universe: C-137
+""".strip()
+    sources = yaml.safe_load(single_source)
+    results = build_source_list(sources)
+
+    assert len(results) == 1
+    assert results[0]["name"] == "source_1"
+    assert results[0]["type"] == "ansible.eda.generic"
+    assert results[0]["args"].keys() == {"event_delay", "payload"}
+    assert results[0].keys() == {"name", "type", "args"}
+
+
+def test_single_source_without_name():
+    single_source = """
+---
+- name: Test sample 001
+  hosts: all
+  sources:
+    - ansible.eda.range:
+        limit: 5
+        delay: 1
+      filters:
+        - noop:
+""".strip()
+
+    sources = yaml.safe_load(single_source)
+    results = build_source_list(sources)
+
+    assert len(results) == 1
+    assert results[0]["name"] == "source_1"
+    assert results[0]["type"] == "ansible.eda.range"
+    assert results[0]["args"] == {"limit": 5, "delay": 1}
+    assert results[0].keys() == {"name", "type", "args"}
+
+
+def test_multiple_sources():
+    sources = """
+---
+- name: Test sample 001
+  hosts: all
+  sources:
+    - name: my source
+      ansible.eda.range1:
+        limit: 5
+        delay: 1
+      filters:
+        - noop:
+    - name: my 2nd source
+      ansible.eda.range2:
+        limit: 6
+        delay: 2
+      filters:
+        - noop:
+""".strip()
+
+    sources = yaml.safe_load(sources)
+    results = build_source_list(sources)
+
+    assert len(results) == 2
+    assert results[0]["name"] == "my source"
+    assert results[0]["type"] == "ansible.eda.range1"
+    assert results[0]["args"] == {"limit": 5, "delay": 1}
+    assert results[1]["name"] == "my 2nd source"
+    assert results[1]["type"] == "ansible.eda.range2"
+    assert results[1]["args"] == {"limit": 6, "delay": 2}
+
+
+def test_multiple_sources_without_names():
+    sources = """
+---
+- name: Test sample 001
+  hosts: all
+  sources:
+    - ansible.eda.range:
+        limit: 5
+        delay: 1
+      filters:
+        - noop:
+    - ansible.eda.range:
+        limit: 6
+        delay: 2
+      filters:
+        - noop:
+""".strip()
+
+    sources = yaml.safe_load(sources)
+    results = build_source_list(sources)
+
+    assert len(results) == 2
+    assert results[0]["name"] == "source_1"
+    assert results[1]["name"] == "source_2"
+
+
+def test_duplicate_sources():
+    sources = """
+---
+- name: Test sample 001
+  hosts: all
+  sources:
+    - name: my source
+      ansible.eda.range:
+        limit: 5
+        delay: 1
+      filters:
+        - noop:
+    - name: my source
```

Comment:
@hsong-rh In this scenario we have the same named source which we should not allow but since we have never enforced we might want to just change the name to be my source #2 since its a duplicate.

Reply:
Blue-skying... maybe when generating a name for a source that does not provide one that we generate the name based on the type? For example the name (for otherwise not named) ansible.eda.range would be ansibile.eda.range-1, ansible.eda.range-2, etc. The -<number> would be per-type so each type would counted separately.

Could be seen as redundant w/ the type for those sources.